snappy-core/src/main/scala/org/apache/spark/sql/store/CodeGeneration.scala

```diff
@@ -0,0 +1,261 @@
+/*
+ * Copyright (c) 2016 SnappyData, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you
+ * may not use this file except in compliance with the License. You
+ * may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License. See accompanying
+ * LICENSE file.
+ */
+package org.apache.spark.sql.store
+
+import java.sql.PreparedStatement
+import java.util.Collections
+
+import scala.collection.JavaConverters._
+
+import com.google.common.cache.{CacheBuilder, CacheLoader}
+import org.codehaus.janino.CompilerFactory
+
+import org.apache.spark.Logging
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.codegen.{BufferHolder, CodeGenContext, GenerateUnsafeProjection}
+import org.apache.spark.sql.catalyst.util.{ArrayData, MapData}
+import org.apache.spark.sql.collection.WrappedRow
+import org.apache.spark.sql.columnar.ExternalStoreUtils
+import org.apache.spark.sql.jdbc.JdbcDialect
+import org.apache.spark.sql.types._
+
+/**
+ * Utilities to generate code for exchanging data from Spark layer
+ * (Row, InternalRow) to store (Statement, ExecRow).
+ * <p>
+ * This is both more efficient and allows better code reuse with the code
+ * generation facilities of Spark (esp for complex types like ArrayType).
+ */
+trait CodeGeneration {
+
+  @throws[java.sql.SQLException]
+  def executeStatement(stmt: PreparedStatement, multipleRows: Boolean,
+      rows: java.util.Iterator[InternalRow], batchSize: Int,
+      schema: Array[StructField], dialect: JdbcDialect): Int
+}
+
+private final class ExecuteKey(val name: String,
+    val schema: Array[StructField], val dialect: JdbcDialect) {
+
+  override def hashCode(): Int = name.hashCode
+
+  override def equals(other: Any): Boolean = other match {
+    case o: ExecuteKey => name == o.name
+    case s: String => name == s
+    case _ => false
+  }
+}
+
+object CodeGeneration extends Logging {
```

Comment:
Yes, Spark's current way will generate code separately for every plan creation. This takes quite a bit of time so didn't want to do that and instead let it be per table based. Can change to be plan based for SnappyContext.update if we are not worried about the performance of that.

Ideally we should be going the whole way and caching the whole plan (with code generation) per PreparedStatement (or statement with tokenization) like done in snappy-store. Then it also requires a mechanism to invalidate plans (we added some start of that already in DependencyCatalog). It will be a good idea to contribute back to Spark but will require more work for proper plan caching which is what we should be doing ultimately.